Uchu.World/Objects/Components/ReplicaComponents/RacingControlComponent.cs

```diff
@@ -1,21 +1,713 @@
+                existingAllTime.NumPlayed++;
+                existingAllTime.Time = Math.Min(existingAllTime.Time, raceTime);
+                existingAllTime.BestLapTime = Math.Min(existingAllTime.BestLapTime, lapTime);
+                existingAllTime.Wins += rank == 1 ? 1 : 0;
+                existingAllTime.LastPlayed = DateTimeOffset.Now.ToUnixTimeSeconds();
+
+                Logger.Debug("AllTime: " + existingAllTime.Time + " Lap: " + existingAllTime.BestLapTime);
+                ctx.ActivityScores.Update(existingAllTime);
+            }
+            // Add new entry.
+            else
+            {
+                ctx.ActivityScores.Add(new ActivityScore
+                {
+                    Activity = this.ActivityInfo.ActivityID ?? 0,
+                    Zone = player.Zone.ZoneId,
+                    CharacterId = player.Id,
+                    Time = raceTime,
+                    BestLapTime = lapTime,
+                    Wins = rank == 1 ? 1 : 0,
+                    LastPlayed = DateTimeOffset.Now.ToUnixTimeSeconds(),
+                    NumPlayed = 1,
+                    Week = 0,
+                });
+            }
+
+            ctx.SaveChanges();
+        }
+
         public override void Construct(BitWriter writer)
         {
-            Serialize(writer);
+            this.Serialize(writer);
         }
 
+        // override to be able to use ScriptedActivityComponent as base
         public override void Serialize(BitWriter writer)
         {
+            for (var i = 0; i < this.Parameters.Count; i++)
+            {
+                var parameters = Parameters[i];
+                parameters[0] = 1;
+                parameters[1] = 265f; // doesn't appear on client
+                parameters[2] = 87f; // doesn't appear on client
+            }
+
             base.Serialize(writer);
+            StructPacketParser.WritePacket(this.GetSerializePacket(), writer);
+        }
+
+        public new RacingControlSerialization GetSerializePacket()
+        {
+            var packet = this.GetPacket<RacingControlSerialization>();
+
+            packet.ExpectedPlayerCount = (ushort)this.Participants.Count;
+
+            packet.PreRacePlayerInfos = this._players.Select(info => new PreRacePlayerInfo
+            {
+                Player = info.Player,
+                Vehicle = info.Vehicle,
+                StartingPosition = info.PlayerIndex,
+                IsReady = info.PlayerLoaded,
+            }).ToArray();
 
-            // TODO: Look into to
+            packet.RaceInfo = _raceInfo;
+
+            packet.DuringRacePlayerInfos = this._players.Select(info => new DuringRacePlayerInfo
+            {
+                Player = info.Player,
+                BestLapTime = (float) info.BestLapTime.TotalSeconds,
+                RaceTime = (float) info.RaceTime.Elapsed.TotalSeconds,
+            }).ToArray();
+
+            // TODO
+
+            return packet;
+        }
+
+        private enum RacingStatus {
```

Comment:
I assume these classes will be moved out like the other packets at some point?

Reply:
I'm not sure about this one. This is supposed to be an internal thing. It has no use outside of RacingControlComponent iirc